keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,473 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The performance of kube-proxy's iptables mode in large clusters could
+be greatly increased by optimizing its calls to `iptables-restore`.
+Although this is a small change in terms of code size, it has a large
+effect on kube-proxy operation, and could potentially completely break
+clusters (if the code was buggy), so we have decided to treat it as a
+KEP-able feature.
+
+## Motivation
+
+### Goals
+
+- Improve the performance of kube-proxy in iptables mode, by optimizing
+  its calls to `iptables-restore` by not including rules that haven't
+  changed.
+
+### Non-Goals
+
+- Everything else.
+
+## Proposal
+
+### General Plan
+
+`iptables-restore` has the constraint that if you want to specify
+_any_ rule in a given iptables chain, then you must specify _every_
+rule in that chain. However, if you want to leave a chain completely
+unchanged, then you can just fail to mention any rules in that chain,
+and it will be left untouched. (If you want to delete a chain you have
+to explicitly request that.)
+
+Thus, [the proposed PR] keeps track of which Service and EndpointSlice
+objects have changed since the last `iptables-restore`, and only
+outputs the Service-specific chains for a Service (the `KUBE-SVC-*`,
+`KUBE-SVL-*`, `KUBE-EXT-*`, `KUBE-FW-*`, and `KUBE-SEP-*` chains) if
+the Service or its EndpointSlice(s) changed (or were deleted) since the last
+`iptables-restore`. (The `KUBE-SERVICES`, `KUBE-EXTERNAL-SERVICES`,
+and `KUBE-NODEPORTS` chains are written in their entirety on every
+sync.)
+
+In theory this could be optimized further; eg, there are situations
+where only `KUBE-FW-*` changes, or where we need to rewrite the
+`KUBE-SERVICES` rules for a service but not any of the
+service-specific chains, or vice versa. However, these other
+optimizations would be more complicated, and would only result in a
+small additional performance improvement relative to the main
+improvement of not rewriting _every_ service and endpoint rule on
+_every_ resync.
+
+In the event that a partial `iptables-restore` call fails, kube-proxy
+will queue another resync, and fall back to doing a full
+`iptables-restore` that time. This should only happen if there is a
+bug in kube-proxy that makes it accidentally write out an inconsistent
+partial ruleset; eg, if kube-proxy mistakenly believed that it had
+previously successfully created the `KUBE-SVC-ABCD` chain, and then
+later tried to restore a ruleset including a `KUBE-SERVICES` rule that
+jumps to `KUBE-SVC-ABCD`, then that restore would fail because of the
+dangling jump. Falling back to a full restore would then fix things,
+because in that case kube-proxy would recreate every chain that it
+needed, including the missing `KUBE-SVC-ABCD`.
+
+Additionally, kube-proxy will always do a full resync when there are
+topology-related changes to Node labels, and it will always do a full
+resync at least once every `iptablesSyncPeriod`.
```

Comment:
Yes, it's mostly paranoia. But as someone else suggested above, we can experiment with changing the default sync period later too...

Reply:
> Yes, it's mostly paranoia. But as someone else suggested above, we can experiment with changing the default sync period later too...

What do you think about adding a metric for "full resync being different than partial resync"?

I'm happy to be proven wrong, but I have a feeling that it may not be that hard because:
1) we will implement the logic for doing partial updates anyway
2) we already have logic for full resyncs anyway

So they way I would do that would be:
- when "full restore is needed" then:
(a) do partial restore
(b) compute "full restore"
(c) "run iptables-save"
(d) compare if what we computed in (b) is the same what we got from (c):
  - if yes - we're done (even without this expensive full restore)
  - if not - do full "iptables restore" and update the metric I'm proposing

Does that make sense? Anything I'm missing?

Reply:
> I'm happy to be proven wrong, but I have a feeling that it may not be that hard because:
> 
> 1. we will implement the logic for doing partial updates anyway
> 2. we already have logic for full resyncs anyway

They're not separate pieces of code though; the logic for doing partial updates is a subversion of the logic for full resyncs. So we'd have to reorganize the code further to be able to do both at once, and that reorganization would be larger than the existing patch to implement partial syncs, thus giving us another chance to introduce new bugs...

(ISTM that the cleanest way to do this would be to split out the rule-generating core of `syncProxyRules` into a helper and call it twice, once asking for a partial sync and once asking for a full sync.)

> * when "full restore is needed" then:

Rather, "when we were going to do a full restore out of paranoia, but not when we were going to do a full restore because it was semantically necessary" (eg, because topology changed).

>   (a) do partial restore
>   (b) compute "full restore"
>   (c) "run iptables-save"
>   (d) compare if what we computed in (b) is the same what we got from (c):

This is unfortunately not simple; the output is expected to be _semantically_ the same as what we input, but it's not guaranteed to be _syntactically_ the same. (eg, values may be canonicalized, the subclauses of the rules may be reordered, etc). And different versions of the iptables binary may rewrite rules differently.

So such a comparison would be hard to get right, and the code to do it would be larger than the PR that it was checking... making it overwhelmingly likely that if it did report that a sync had failed, that the bug would actually turn out to have been in the checking code, not in the actual sync...

----

The current version of the KEP has a Beta graduation criterion of 

>     - Real-world users trying out the Alpha feature have reported both
>       (a) no new bugs, and (b) improved performance.

I am happy to let people be as draconian with the interpretation of that criterion as they want to be. I think getting real-world evidence is just going to be much easier than trying to write code to prove the code correct.

(I already have users waiting to test this code. It's possible some of them will already be using it production before 1.26 ships...)

Reply:
> I think getting real-world evidence is just going to be much easier than trying to write code to prove the code correct.

I'm not convinced. If e.g. in a real world we have a service with 1000 endpoints and our partial update logic will miss to add a new one - there is no way someone will really notice it, especially given it will self-restore on full sync a bit later.
This is the kind of bugs that I would really like to avoid, because it would effectively degrade the network programming latency in case of hitting the buggy branch. 


Now, I fully understand that it would require (much) larger refactoring that was planned, because the code would need to be structured differently to making comparison simpler. It seems that you're not necessary disagreeing with this:

> (ISTM that the cleanest way to do this would be to split out the rule-generating core of syncProxyRules into a helper and call it twice, once asking for a partial sync and once asking for a full sync.)

And wouldn't that refactor actually be desired then?


> This is unfortunately not simple; the output is expected to be semantically the same as what we input, but it's not guaranteed to be syntactically the same. (eg, values may be canonicalized, the subclauses of the rules may be reordered, etc). And different versions of the iptables binary may rewrite rules differently.

Hmm - I think I'm not convinced. Things like "different iptables binaries" would happen across kube-proxy restarts anyway, so I don't think it's a problem (on restart we shouldn't be reporting differences anyway).

I didn't touch this code for quite some time, so you're probably right that some reorderings and/or canonicalizations may happen, but making the code addressing the ordering issue shouldn't be hard

And finally - even if we wouldn't address 100% of comparison issues, I would prefer to have a metric that says: "there is a possibility that our partial update didn't do its job, because the syntax has changed - please check the logs to see if this isn't a false positive" that not having nothing.


I don't want to block Alpha on this with upcoming code freeze, but can we maybe put a Beta graduation criteria to make a decision whether we are really sure we want to skip having such metric and sacrifice the potential benefits of having it?
@danwinship ^^

Also @thockin - for his thoughts.

Reply:
What if we would say that we implement such comparison code but keep it disabled by default (to avoid confusing users), but enable it in our tests?